virtualizarr/readers/common.py

```diff
@@ -1,121 +1,107 @@
-from abc import ABC
-from collections.abc import Iterable, Mapping, MutableMapping
+from collections.abc import Iterable, Mapping
 from typing import (
     Any,
     Hashable,
+    MutableMapping,
     Optional,
 )
 
-import xarray  # noqa
-from xarray import (
-    Coordinates,
-    Dataset,
-    DataTree,
-    Index,
-    IndexVariable,
-    Variable,
-    open_dataset,
-)
-from xarray.core.indexes import PandasIndex
+import xarray as xr
+import xarray.indexes
 
 from virtualizarr.utils import _FsspecFSFromFilepath
 
 
-def maybe_open_loadable_vars_and_indexes(
-    filepath: str,
-    loadable_variables,
-    reader_options,
-    drop_variables,
-    indexes,
-    group,
-    decode_times,
-) -> tuple[Mapping[str, Variable], Mapping[str, Index]]:
-    """
-    Open selected variables and indexes using xarray.
+def construct_fully_virtual_dataset(
+    virtual_vars: Mapping[str, xr.Variable],
+    coord_names: Iterable[str] | None = None,
+    attrs: dict[str, Any] | None = None,
+) -> xr.Dataset:
+    """Construct a fully virtual Dataset from constituent parts."""
+
+    data_vars, coords = separate_coords(
+        vars=virtual_vars,
+        indexes={},  # we specifically avoid creating any indexes yet to avoid loading any data
+        coord_names=coord_names,
+    )
 
-    Relies on xr.open_dataset and its auto-detection of filetypes to find the correct installed backend.
-    """
+    vds = xr.Dataset(
+        data_vars=data_vars,
+        coords=coords,
+        attrs=attrs,
+    )
 
-    if loadable_variables == [] and indexes == {}:
-        # no need to even attempt to open the file using xarray
-        return {}, {}
+    # TODO we should probably also use vds.set_close() to tell xarray how to close the file we opened
 
-    # TODO We are reading a bunch of stuff we know we won't need here, e.g. all of the data variables...
-    # TODO It would also be nice if we could somehow consolidate this with the reading of the kerchunk references
-    # TODO Really we probably want a dedicated backend that iterates over all variables only once
-    # TODO See issue #124 for a suggestion of how to avoid calling xarray here.
+    return vds
+
+
+# TODO reimplement this using ManifestStore (GH #473)
+def replace_virtual_with_loadable_vars(
+    fully_virtual_dataset: xr.Dataset,
+    filepath: str,  # TODO won't need this after #473 because the filepaths will be in the ManifestStore
+    group: str | None = None,
+    loadable_variables: Iterable[Hashable] | None = None,
```

Comment:
Why is the type `Iterable[Hashable]` here rather than `Iterable[str]` like elsewhere?

Reply:
Because technically xarray variable names can be `Hashable`. It should probably be generalized elsewhere too. I think I changed it here just to get rid of a mypy error.